Report/final_report.tex

```diff
@@ -277,33 +293,69 @@
 
-    We also removed some country names because they were not present in both datasets, making the actual data presented in our application even more limited.
+        Then, we switched to \verb|QThread| which is preferred by \verb|PyQt5|. However, another problem raised that as we know, \verb|PyQt5| is thread-safe which means that it doesn't allow any modifications from another \verb|QThread|, so we could not simply run initialization on another \verb|QThread| and directly report the progress to the \verb|PyQt5|.
 
-    \subsubsection{Obstacles}
+        When we nearly gave up implementing the progress bar, the angels came. We remembered the \verb|PyQt5| had the signals and slots mechanism (we explained earlier), and we could actually create a new signal in our monitor thread and send this signal to the main gui thread every time the progress updated. In this way, we utilized the built-in mechanism of \verb|PyQt5| and prevented any thread-unsafe operations. 
 
-    Our application crashes all day, all night, 24 * 7, non-stop. Making us want to drop out of CSC110. But nevertheless, we still overcame the difficulties and produced a "Functioning" application.
+        Retrospectively, because we learned many valuable knowledge from this challenge and now we know how to tackle it, the progress bar is not very difficult. However, although the progress bar itself is trivial, the processes involved in overcoming the challenge are pivotal to us. 
 
-    \subsection{Next steps}
+        \newpage
 
-    We will be carrying the skills we developed in this project into other projects in the future. We have now acquired the ability to download, process, and hash data. These abilities would be necessary to create any automated application in the future.
+        \item The other challenge is about optimizations.
+        
+        In our plot, we implemented a cross-hair that centered on the closet point in the line from the cursor. We successfully implemented it with binary search and many data transformations and calculations. However, the frame rate per second (FPS) was very low when moving the cross-hair.
 
-    We also acquired one of the most important skills in the Computer Science Industry, which is to design a Graphical User Interface. This is the ultimate version of "Designer to User" contract, which provides a very "Easy to use" interface so that users without any knowledge in Computer Science can also use this application to their advantage.
+        To investigate the root of the problem, we analyzed the running time of our algorithms but it turned out that the running time was good ($\bigO(n)$). 
```

Comment:
> ...we analyzed the running time of our algorithms but it turned out that the running time was good ($\bigO(n)$). 

...we analyzed the running time of our algorithms, but it turned out that the running time was fairly decent ($\bigO(n)$).